draft-ietf-quic-tls.md

```diff
@@ -499,12 +499,24 @@ handshake, new data is requested from TLS after providing received data.
 ### Encryption Level Changes
 
 As keys at a given encryption level become available to TLS, TLS indicates to
-QUIC that reading or writing keys at that encryption level are available.  While
-generating these keys, an endpoint SHOULD buffer received packets marked as
-protected by the keys being generated, and process them once those keys become
-available.  If the keys are generated asynchronously, an endpoint MAY continue
-responding to the received packets that were processable while waiting for TLS
-to provide these keys.
+QUIC that reading or writing keys at that encryption level are available.
+
+The events that cause new keys to be available are not asynchronous; they
+always occur immediately after TLS is provided with inputs. Inputs are either
+new handshake bytes or new instructions. The two instructions that this
+document relies upon are the initial signal to start the handshake and - if the
+TLS implementation depends on certificate validation being performed externally
+- an indication that the certificate chain of a peer has been accepted or
+rejected.
```

Comment:
As stated above, I have a mixed feeling about this paragraph. Defining the inputs to TLS stack is clean, but it is complex. Generally speaking, I think it is a good idea to treat TLS as a black box whenever possible.

To make a suggestion, how about changing this paragraph to something like: _Some TLS stacks might provide new keys as QUIC provides TLS the new handshake bytes. Others might provide those keys at a later point, for example due to certificate validation done in an asynchronous manner._

Reply:
Let me see what that looks like.  I agree that this external validation thing is weird and probably needs work.

Reply:
Thanks for persisting @kazuho, I think that worked out better.